src/unix/futex.c

```diff
@@ -294,3 +294,103 @@ init_futices(process p)
         halt("failed to allocate futex table\n");
 
 }
+
+/* robust mutex handling */
+
+#define FUTEX_OWNER_DIED	0x40000000
+#define FUTEX_KEY_ADDR(x, o)    ((int *)((u8 *)(x) + (o)))
+
+typedef struct robust_list {
+    struct robust_list *next;
+} *robust_list;
+
+typedef struct robust_list_head {
+    robust_list list;
+    long futex_offset;
+    void *list_op_pending;
+} *robust_list_head;
+
+static boolean valid_user_address(process p, void *a, int len)
+{
+    u64 pg = u64_from_pointer(a) & ~PAGEMASK;
+    u64 end = (u64_from_pointer(a) + len) & ~PAGEMASK;
+
+    for (; pg <= end; pg += PAGESIZE) {
+        if (vmap_from_vaddr(p, pg) == INVALID_ADDRESS)
+                return false;
+    }
+    return true;
+}
+
+void wake_robust_list(process p, void *head)
+{
+    struct robust_list_head *h = head;
+    struct robust_list *l;
+    int *uaddr;
+
+    /* must be very careful accessing the head as well as the list */
+    if (!valid_user_address(p, h, sizeof *h))
```

Comment:
A stylistic nit, but elsewhere in the codebase we've been using parentheses with sizeof.